src/terminator/mod.rs

```diff
@@ -367,31 +367,43 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
 
         // Only trait methods can have a Self parameter.
 
+        // Intercept some methods (even if we can find MIR for them)
+        if let ty::InstanceDef::Item(def_id) = instance.def {
+            match &self.tcx.item_path_str(def_id)[..] {
+                "std::sys::imp::fast_thread_local::register_dtor" => {
```

Comment:
Oh I was misremembering, this is easier than I thought (it also proves that the semantics are tricky)! Turns out [the `static` has to contain a pointer](https://github.com/rust-lang/rust/blob/deb90c33ccee849821de5f3a8e26d8d9a2e774e5/src/librustc_trans/consts.rs#L128) which can be observed as null (unless overridden). So you were on the right track, just need to bypass `static` initializer evaluation.